### ISSUE TITLE
[Consensus] Include the consensus time into block interval.

### ISSUE THREAD

**Issue by Jim8y** [neo-project/neo#3627]
**Summary or problem description**
Currently we *wait `block interval` seconds `after receiving the previous block` before starting a new round of consensus*, this excludes the consensus time from the `block interval` and makes the actual `block time` longer than the `block interval`. With an expected of `block time` as `block interval` seconds, the actual `block time` is always larger than that, 1-2 seconds and more.

**Do you have any solution you want to propose?**
We can include the consensus time into the `block interval` by *waiting `block interval` seconds `after receiving the previous consensus reparerequest` before starting a new round of consensus*.  This way, as long as the consensus delay is less than the `block interval` seconds, we can ensure a fixed `block time`, with minor deviation. 

<img width="823" alt="image" src="https://github.com/user-attachments/assets/20cc53c9-9adc-475a-9b61-654f7fbcb7d0" />

**Where in the software does this update applies to?**
- Consensus

**Comment by Jim8y** [comment]
This is the implementation: https://github.com/neo-project/neo/pull/3622

**Comment by roman-khimov** [comment]
Hi, https://github.com/nspcc-dev/dbft/issues/55, https://github.com/nspcc-dev/dbft/pull/56.

**Comment by Jim8y** [comment]
> > This is the implementation: #3622
> 
> Please create a PR just with this change.

@vncoelho it will not be applied to current 15 seconds consensus, but be considered to be applied to 3 seconds solution.

**Comment by shargon** [comment]
> > This is the implementation: #3622
> 
> Please create a PR just with this change.

Agree

**Comment by Jim8y** [comment]
> > > This is the implementation: #3622
> > 
> > 
> > Please create a PR just with this change.
> 
> Agree

Can't be, create a seperate pr only contain this change means it will be applied to current system, while this will increase the gas distrubution by 10-12%. It should only be applied to the 3 seconds consensus.

**Comment by roman-khimov** [comment]
> It should only be applied to the 3 seconds consensus

This makes no sense to me. All of the current GAS distribution assumes 15s blocks by design. The fact that some blocks deviate from these 15s is certainly not a feature that needs to be kept. The lower the delay the better, absolutely irrespective of block time.

**Comment by Jim8y** [comment]
> > It should only be applied to the 3 seconds consensus
> 
> This makes no sense to me. All of the current GAS distribution assumes 15s blocks by design. The fact that some blocks deviate from these 15s is certainly not a feature that needs to be kept. The lower the delay the better, absolutely irrespective of block time.

Yes, but vitor is claming that 3-seconds can slow the gas distribution already, so what i want to focus is only how to minimize the impact to gas distribution in 3-seconds consenus, instead of fixing current one. Yes, currenly it assumes 15 seconds, but as you also know it never happened, the average block time for mainnet is around 16.8 seconds recently, so applying this will directly cause 10% more gas being minted, even if its actually expected, not a thing that i, just personally,  want to touch.

**Comment by roman-khimov** [comment]
Have you tried running the same 3s network _without_ this change and with the same tests?

**Comment by Jim8y** [comment]
> Have you tried running the same 3s network _without_ this change and with the same tests?

No, cause just as what we have observed in the mainnet, consensus itself takes 1-2 seconds, so, for me, running that test is meanless due to high delay.

**Comment by roman-khimov** [comment]
You can't rely on mainnet data for this, mainnet is a very different network.

**Comment by Jim8y** [comment]
> You can't rely on mainnet data for this, mainnet is a very different network.

for dbft they are the same, yes, they may have minor difference, but that does not affect the dbft process itself. Like it takes 1-3 seconds to run the whole dbft steps, it makes no sense it takes less time in 3-seconds-consensus. I understand that it is best to do it to make sure.....but it is not in a hurry, right.

**Comment by roman-khimov** [comment]
If you want to have any representative results for #3637 this is just the way to test it: take a network (whatever that is), run it without a patch (with/without load), collect metrics, run _the same_ network with a patch with the same tests and then compare the result. There is no other way to do it if you care about real results. Because

> it takes 1-3 seconds to run the whole dbft steps in mainnet

is read properly as "we don't know how long it takes on our test network".

Mainnet is estimated as 100+ nodes network. Mainnet has different node implementations and versions including ones that are completely forgotten, stale and misbehaving. Comparing things happening there with things going on in a network of seven uniform fresh nodes directly is not correct.
